File: src/main/java/com/telenordigital/prime/firebase/FbStorage.java
package com.telenordigital.prime.firebase;

import com.google.firebase.FirebaseApp;
import com.google.firebase.FirebaseOptions;
import com.google.firebase.auth.FirebaseCredentials;
import com.google.firebase.database.FirebaseDatabase;
import com.telenordigital.prime.events.EventListeners;
import com.telenordigital.prime.storage.ProductDescriptionCache;
import com.telenordigital.prime.storage.ProductDescriptionCacheImpl;
import com.telenordigital.prime.storage.PurchaseRequestListener;
import com.telenordigital.prime.storage.Storage;
import com.telenordigital.prime.storage.StorageException;
import com.telenordigital.prime.storage.entities.Product;
import com.telenordigital.prime.storage.entities.PurchaseRequest;
import com.telenordigital.prime.storage.entities.Subscriber;
import com.telenordigital.prime.storage.entities.SubscriberImpl;

import java.io.FileInputStream;
import java.io.IOException;
import java.util.Collection;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

public final class FbStorage implements Storage {

    private final ProductDescriptionCache productCache;

    private final FbDatabaseFacade facade;

    private final EventListeners listeners;

    public FbStorage(final String databaseName,
                     final String configFile,
                     final EventListeners listeners) throws StorageException {

        checkNotNull(configFile);
        checkNotNull(databaseName);
        this.listeners = checkNotNull(listeners);

        this.productCache = ProductDescriptionCacheImpl.getInstance();

        final FirebaseDatabase firebaseDatabase;
        firebaseDatabase = setupFirebaseInstance(databaseName, configFile);

        this.facade = new FbDatabaseFacade(firebaseDatabase);

        facade.addProductCatalogItemListener(listeners::productCatalogItemListener);
        facade.addPurchaseRequestListener(listeners::purchaseRequestListener);

        // Load subscriber balance from firebase to in-memory OcsState
        listeners.loadSubscriberBalanceDataFromFirebaseToInMemoryStructure(getAllSubscribers());
    }

    @Override
    public void addTopupProduct(final String sku, final long noOfBytes) {
        productCache.addTopupProduct(sku, noOfBytes);
    }

    @Override
    public boolean isValidSKU(final String sku) {
        return productCache.isValidSKU(sku);
    }

    @Override
    public Product getProductForSku(final String sku) {
        return productCache.getProductForSku(sku);
    }

    private FirebaseDatabase setupFirebaseInstance(
            final String databaseName,
            final String configFile) throws StorageException {
        try (FileInputStream serviceAccount = new FileInputStream(configFile)) {

            final FirebaseOptions options = new FirebaseOptions.Builder().
                    setCredential(FirebaseCredentials.fromCertificate(serviceAccount)).
                    setDatabaseUrl("https://" + databaseName + ".firebaseio.com/").
                    build();

            try {
                FirebaseApp.getInstance();
            } catch (Exception e) {
                FirebaseApp.initializeApp(options);
            }

            return FirebaseDatabase.getInstance();

            // (un)comment next line to turn on/of extended debugging
            // from firebase.
            // this.firebaseDatabase.setLogLevel(com.google.firebase.database.Logger.Level.DEBUG);

        } catch (IOException ex) {
            throw new StorageException(ex);
        }
    }


    // XXX This method represents a bad design decision.  It's too circumspect to
    //     understand.  Fix!
    @Override
    public void addPurchaseRequestListener(final PurchaseRequestListener listener) {
        checkNotNull(listener);
        listeners.addPurchaseRequestListener(listener);
    }

    @Override
    public void updateDisplayDatastructure(final String msisdn) throws StorageException {
        checkNotNull(msisdn);
        final SubscriberImpl subscriber = (SubscriberImpl) getSubscriberFromMsisdn(msisdn);

        if (subscriber == null) {
            throw new StorageException("Unknown MSISDN " + msisdn);
        }

        final long noOfBytes = subscriber.getNoOfBytesLeft();
        final float noOfGBLeft = noOfBytes / 1.0E09f;
        final String gbLeft = String.format("%.2f GB", noOfGBLeft);

        facade.updateClientVisibleUsageString(msisdn, gbLeft);
    }

    @Override
    public void removeDisplayDatastructure(final String msisdn) throws StorageException {
        checkNotNull(msisdn);
        facade.removeByMsisdn(msisdn);
    }

    @Override
    public String injectPurchaseRequest(final PurchaseRequest pr) {
        checkNotNull(pr);
        return facade.injectPurchaseRequest(pr);
    }

    @Override
    public void removeRecordOfPurchaseById(final String id) {
        checkNotNull(id);
        facade.removeRecordOfPurchaseById(id);
    }

    @Override
    public String addRecordOfPurchaseByMsisdn(
            final String msisdn,
            final String sku,
            final long millisSinceEpoch) throws StorageException {
        checkNotNull(msisdn);
        checkNotNull(sku);
        checkArgument(millisSinceEpoch > 0);

        return facade.addRecordOfPurchaseByMsisdn(msisdn, sku, millisSinceEpoch);
    }

    @Override
    public void removeSubscriberByMsisdn(final String msisdn) throws StorageException {
        checkNotNull(msisdn);
        facade.removeSubscriberByMsisdn(msisdn);
    }

    @Override
    public void removePurchaseRequestById(final String id) {
        checkNotNull(id);
        facade.removePurchaseRequestById(id);
    }

    @Override
    public Subscriber getSubscriberFromMsisdn(final String msisdn) throws StorageException {
        checkNotNull(msisdn);
        return facade.getSubscriberFromMsisdn(msisdn);
    }

    @Override
    public void setRemainingByMsisdn(
            final String msisdn,
            final long noOfBytes) throws StorageException {

        if (msisdn == null) {
            throw new StorageException("msisdn can't be null");
        }

        if (noOfBytes < 0) {
            throw new StorageException("noOfBytes can't be negative");
        }

        final SubscriberImpl sub = (SubscriberImpl) getSubscriberFromMsisdn(msisdn);

        if (sub == null) {
            throw new StorageException("Unknown msisdn " + msisdn);
        }

        sub.setNoOfBytesLeft(noOfBytes);

        facade.updateAuthorativeUserData(sub);
    }

    @Override
    public String insertNewSubscriber(final String msisdn) {
        checkNotNull(msisdn);
        final SubscriberImpl sub = new SubscriberImpl();
        sub.setMsisdn(msisdn);
        return facade.insertNewSubscriber(sub);
    }

    @Override
    public Collection<Subscriber> getAllSubscribers() {
        return facade.getAllSubscribers();
    }
}

File: prime/src/main/kotlin/org/ostelco/prime/firebase/FbStorage.kt
package org.ostelco.prime.firebase

import com.google.auth.oauth2.GoogleCredentials
import com.google.common.base.Preconditions.checkArgument
import com.google.common.base.Preconditions.checkNotNull
import com.google.firebase.FirebaseApp
import com.google.firebase.FirebaseOptions
import com.google.firebase.database.FirebaseDatabase
import org.ostelco.prime.events.EventHandler
import org.ostelco.prime.model.Product
import org.ostelco.prime.model.PurchaseRequest
import org.ostelco.prime.model.RecordOfPurchase
import org.ostelco.prime.model.Subscriber
import org.ostelco.prime.storage.ProductDescriptionCache
import org.ostelco.prime.storage.ProductDescriptionCacheImpl
import org.ostelco.prime.storage.PurchaseRequestHandler
import org.ostelco.prime.storage.Storage
import org.ostelco.prime.storage.StorageException
import java.io.FileInputStream
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.util.function.BiFunction
import java.util.function.Consumer

class FbStorage @Throws(StorageException::class)
constructor(databaseName: String,
            configFile: String,
            eventHandler: EventHandler) : Storage {

    private val productCache: ProductDescriptionCache

    private val facade: FbDatabaseFacade

    private val eventHandler: EventHandler

    override val allSubscribers: Collection<Subscriber>
        get() = facade.allSubscribers

    init {

        checkNotNull(configFile)
        checkNotNull(databaseName)
        this.eventHandler = checkNotNull(eventHandler)

        this.productCache = ProductDescriptionCacheImpl

        val firebaseDatabase = setupFirebaseInstance(databaseName, configFile)

        this.facade = FbDatabaseFacade(firebaseDatabase)

        facade.addProductCatalogItemHandler(Consumer { eventHandler.productCatalogItemHandler(it) })
    }

    override fun addTopupProduct(sku: String, noOfBytes: Long) {
        productCache.addTopupProduct(sku, noOfBytes)
    }

    override fun isValidSKU(sku: String): Boolean {
        return productCache.isValidSKU(sku)
    }

    override fun getProductForSku(sku: String): Product? {
        return productCache.getProductForSku(sku)
    }

    @Throws(StorageException::class)
    private fun setupFirebaseInstance(
            databaseName: String,
            configFile: String): FirebaseDatabase {

        try {

            val credentials: GoogleCredentials = if (Files.exists(Paths.get(configFile))) {
                FileInputStream(configFile).use { serviceAccount -> GoogleCredentials.fromStream(serviceAccount) }
            } else {
                GoogleCredentials.getApplicationDefault()
            }

            val options = FirebaseOptions.Builder()
                    .setCredentials(credentials)
                    .setDatabaseUrl("https://$databaseName.firebaseio.com/")
                    .build()
            try {
                FirebaseApp.getInstance()
            } catch (e: Exception) {
                FirebaseApp.initializeApp(options)
            }

            return FirebaseDatabase.getInstance()

            // (un)comment next line to turn on/of extended debugging
            // from firebase.
            // this.firebaseDatabase.setLogLevel(com.google.firebase.database.Logger.Level.DEBUG);
        } catch (ex: IOException) {
            throw StorageException(ex)
        }
    }


    // XXX This method represents a bad design decision.  It's too circumspect to
    //     understand.  Fix!
    override fun addPurchaseRequestHandler(handler: PurchaseRequestHandler) {
        checkNotNull(handler)
        eventHandler.addPurchaseRequestHandler(handler)
        facade.addPurchaseRequestListener(BiFunction { key, req -> eventHandler.purchaseRequestHandler(key, req) })
    }

    @Throws(StorageException::class)
    override fun updateDisplayDatastructure(msisdn: String) {
        checkNotNull(msisdn)
        val subscriber = getSubscriberFromMsisdn(msisdn)
                ?: throw StorageException("Unknown MSISDN " + msisdn)

        val noOfBytes = subscriber.noOfBytesLeft
        val noOfGBLeft = noOfBytes / 1.0E09f
        val gbLeft = String.format("%.2f GB", noOfGBLeft)

        facade.updateClientVisibleUsageString(msisdn, gbLeft)
    }

    @Throws(StorageException::class)
    override fun removeDisplayDatastructure(msisdn: String) {
        checkNotNull(msisdn)
        facade.removeDisplayDatastructureByMsisdn(msisdn)
    }

    override fun injectPurchaseRequest(pr: PurchaseRequest): String {
        checkNotNull(pr)
        return facade.injectPurchaseRequest(pr)
    }

    override fun removeRecordOfPurchaseById(id: String) {
        checkNotNull(id)
        facade.removeRecordOfPurchaseById(id)
    }

    override fun addRecordOfPurchase(purchase: RecordOfPurchase): String {
        checkNotNull(purchase.msisdn)
        checkNotNull(purchase.sku)
        checkArgument(purchase.millisSinceEpoch > 0)

        return facade.addRecordOfPurchase(purchase)
    }

    @Throws(StorageException::class)
    override fun removeSubscriberByMsisdn(msisdn: String) {
        checkNotNull(msisdn)
        facade.removeSubscriberByMsisdn(msisdn)
    }

    override fun removePurchaseRequestById(id: String) {
        checkNotNull(id)
        facade.removePurchaseRequestById(id)
    }

    @Throws(StorageException::class)
    override fun getSubscriberFromMsisdn(msisdn: String): Subscriber? {
        checkNotNull(msisdn)
        return facade.getSubscriberFromMsisdn(msisdn)
    }

    @Throws(StorageException::class)
    override fun setRemainingByMsisdn(
            msisdn: String?,
            noOfBytes: Long) {

        if (msisdn == null) {
            throw StorageException("msisdn can't be null")
        }

        if (noOfBytes < 0) {
            throw StorageException("noOfBytes can't be negative")
        }

        val sub = Subscriber(msisdn, noOfBytes)

        facade.updateAuthorativeUserData(sub)
    }

    override fun insertNewSubscriber(msisdn: String) {
        checkNotNull(msisdn)
        val sub = Subscriber(msisdn, 0)
        return facade.insertNewSubscriber(sub)
    }

    override fun addNotification(subscriber: Subscriber) {
        facade.addNotification(subscriber)
    }
}
